### PR TITLE
feat: add loan collateral component

### PR DESCRIPTION
**Pull request by BLasan**
## Description
Component for creating loan collateral added.

## Related issues and discussion
#831 

## Screenshots, if any
Screenshot from 2020-06-06 17-08-58

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] If you have multiple commits please combine them into one commit by squashing them.

- [x] Read and understood the contribution guidelines at `web-app/.github/CONTRIBUTING.md`.

### CODE REVIEW

**Comment by BLasan** [comment]
@muskankhedia Please review

File: src/environments/environment.ts [hunk 1]
```diff
@@ -10,7 +10,7 @@ export const environment = {
   production: false,
   version: env.npm_package_version + '-dev',
   fineractPlatformTenantId: 'default',  // For connecting to server running elsewhere update the tenant identifier
-  baseApiUrl: 'https://dev.mifos.io',  // For connecting to server running elsewhere update the base API URL
+  baseApiUrl: 'https://demo.mifos.io',  // For connecting to server running elsewhere update the base API URL
```
Comment (muskankhedia) [inline src/environments/environment.ts:13]: Please avoid updating these files

File: src/app/loans/loans-routing.module.ts [hunk 2]
```diff
@@ -5,113 +5,139 @@ import { Routes, RouterModule } from '@angular/router';
 /** Translation Imports */
```
Comment (muskankhedia) [inline src/app/loans/loans-routing.module.ts:5]: The loan module is lazy loaded with the clients component. Please revert the changes, only add the required routing with the previous ones. To know more about it, you can view PR #661.

File: src/app/loans/add-collateral/add-collateral.component.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+   * @param {ActivatedRoute} route Activated Route.
+   * @param {LoansService} LoansService loans service.
+   */
+  constructor(private formBuilder: FormBuilder,
+              private router: Router,
+              private route: ActivatedRoute,
+              private loanService: LoansService ) {
+    this.route.data.subscribe((data: {collateralData: any, loanData: any}) => {
+      this.collateralData = data.collateralData;
+      this.loanId = data.loanData.id;
```
Comment (muskankhedia) [inline src/app/loans/add-collateral/add-collateral.component.ts:39]:
```suggestion
      this.loanId = this.route.parent.snapshot.params['loanId'];
```
You don't need to call the API to get the loan id.

File: src/environments/environment.ts (hunk 1 above)
Comment (BLasan) [reply src/environments/environment.ts:13]: Sorry, my mistake

File: src/app/loans/loans-routing.module.ts (hunk 2 above)
Comment (BLasan) [reply src/app/loans/loans-routing.module.ts:5]: After reverting I'm getting the previous error

Comment (BLasan) [reply src/app/loans/loans-routing.module.ts:5]: Can you upload the screenshot of environment.ts file?

Comment (muskankhedia) [reply src/app/loans/loans-routing.module.ts]: Screenshot from 2020-06-06 16-45-23

Comment (BLasan) [reply src/app/loans/loans-routing.module.ts]: same. But not working

Comment (muskankhedia) [reply src/app/loans/loans-routing.module.ts]: Okay. I am trying in my system.

Till then can you specify the path you are following for loans module.

File: src/app/loans/loans-routing.module.ts [hunk 4]
```diff
@@ -11,22 +11,26 @@ import { LoansViewComponent } from './loans-view/loans-view.component';
...
-import { OriginalScheduleTabComponent } from './loans-view/original-schedule-tab/original-schedule-tab.component';
-import { RepaymentScheduleTabComponent } from './loans-view/repayment-schedule-tab/repayment-schedule-tab.component';
-import { TransactiosTabComponent } from './loans-view/transactios-tab/transactios-tab.component';
+import { LoanCollateralTemplateResolver } from './common-resolvers/collateral-template.resolver';
 
+/** Loans Route. */
 const routes: Routes = [
   {
-    path: '',
+    path: 'loans',
```
Comment (muskankhedia) [inline src/app/loans/loans-routing.module.ts:32]: Please remove this change as well

File: src/app/loans/add-collateral/add-collateral.component.html [hunk 5]
```diff
@@ -0,0 +1,51 @@
+<div class="container">
```
Comment (muskankhedia) [inline src/app/loans/add-collateral/add-collateral.component.html:1]: Add the collateral component within the 'loans-view' folder.

File: src/app/loans/loans-routing.module.ts (hunk 4 above)
Comment (BLasan) [reply src/app/loans/loans-routing.module.ts:32]: This was the mistake for getting that issue

File: src/app/loans/loans-routing.module.ts (hunk 2 above)
Comment (BLasan) [reply src/app/loans/loans-routing.module.ts]: It worked. Thanks :)

**Comment by BLasan** [comment]
@muskankhedia  Please review

**Comment by muskankhedia** [comment]
In order to run to for testing, you can lazy-load the components. And for now, the component is working fine, will merge it as soon as I link the buttons with routes so the CORS error won't prevail.

**Comment by muskankhedia** [comment]
@BLasan Please implement this as well as an action button. in order to avoid confusion you can check the `button.js` file to know all the action buttons.

**Comment by BLasan** [comment]
@muskankhedia Done. Please review

File: src/app/loans/common-resolvers/collateral-template.resolver.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
+/** Angular Imports */
```
Comment (muskankhedia) [inline src/app/loans/common-resolvers/collateral-template.resolver.ts:1]: Add the particular condition in the `loan-action-button.resolver.ts` file.

File: src/app/loans/loans-view/loan-account-actions/add-collateral/add-collateral.component.html [hunk 7]
```diff
@@ -0,0 +1,47 @@
+<div class="container">
+
+  <form [formGroup]="collateralForm" (ngSubmit)="submit()">
+
+    <mat-card-content>
+
+      <div fxLayout="column">
+
+        <mat-form-field>
```
Comment (muskankhedia) [inline src/app/loans/loans-view/loan-account-actions/add-collateral/add-collateral.component.html:9]: This is a required field.

File: src/app/loans/loans-view/loan-account-actions/add-collateral/add-collateral.component.ts [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+/**
+ * Add Collateral component.
+ */
+@Component({
+  selector: 'mifosx-add-collateral',
+  templateUrl: './add-collateral.component.html',
+  styleUrls: ['./add-collateral.component.scss']
+})
+export class AddCollateralComponent implements OnInit {
+
```
Comment (muskankhedia) [inline src/app/loans/loans-view/loan-account-actions/add-collateral/add-collateral.component.ts:18]: Implement `@Input()` to receive the data in the component.

File: src/app/loans/loans-view/loan-account-actions/add-collateral/add-collateral.component.ts [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+  }
+
+  /**
+   * Set Collateral form.
+   */
+  setCollateralForm() {
+    this.collateralForm = this.formBuilder.group({
+      'collateralTypeId': ['', Validators.required],
+      'value': ['', Validators.required],
+      'description': ['', Validators.required]
```
Comment (muskankhedia) [inline src/app/loans/loans-view/loan-account-actions/add-collateral/add-collateral.component.ts:55]: `Description` is an optional field.

File: src/app/loans/loans-view/loan-account-actions/add-collateral/add-collateral.component.html [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+          <input matInput required formControlName="value">
+          <mat-error *ngIf="collateralForm.controls.value.hasError('required')">
+            Value is <strong>required</strong>
+          </mat-error>
+        </mat-form-field>
+
+        <mat-form-field>
+          <mat-label>Description</mat-label>
+          <textarea matInput formControlName="description"></textarea>
+          <mat-error *ngIf="collateralForm.controls.description.hasError('required')">
```
Comment (muskankhedia) [inline src/app/loans/loans-view/loan-account-actions/add-collateral/add-collateral.component.html:30]: **Description** is an optional field.

File: src/app/loans/common-resolvers/collateral-template.resolver.ts (hunk 6 above)
Comment (BLasan) [reply src/app/loans/common-resolvers/collateral-template.resolver.ts:1]: Collaterals template parameters are different. Can I use different method for getting collateral template?

Comment (muskankhedia) [reply src/app/loans/common-resolvers/collateral-template.resolver.ts:1]: Yaa sure.

File: src/app/loans/loans-view/loan-account-actions/add-collateral/add-collateral.component.html (hunk 7 above)
Comment (BLasan) [reply src/app/loans/loans-view/loan-account-actions/add-collateral/add-collateral.component.html:9]: Done. Please check

File: src/app/loans/loans-view/loan-account-actions/add-collateral/add-collateral.component.ts (hunk 8 above)
Comment (BLasan) [reply src/app/loans/loans-view/loan-account-actions/add-collateral/add-collateral.component.ts:18]: Done

File: src/app/loans/loans.service.ts [hunk 11]
```diff
@@ -50,10 +50,32 @@ export class LoansService {
...
 
+  /**
+   * Get collateral template.
+   * @param {string} loanId Loan Id.
+   * @returns {Observable<any>}
+   */
+  getLoanCollateralTemplate(loanId: string): Observable<any> {
+    const httpParams = new HttpParams()
+      .set('tenantidentifier', 'default')
+      .set('pretty', 'true');
```
Comment (muskankhedia) [inline src/app/loans/loans.service.ts:65]:
```suggestion
```

I don't think this is required.

File: src/app/shared/shared.module.ts [hunk 12]
```diff
@@ -29,7 +29,7 @@ import { MaterialModule } from './material.module';
     CommonModule,
     IconsModule,
     MaterialModule,
-    ReactiveFormsModule
+    ReactiveFormsModule,
```
Comment (muskankhedia) [inline src/app/shared/shared.module.ts:32]: Please revert this change

File: src/app/app.module.ts [hunk 13]
```diff
@@ -39,6 +39,7 @@ import { ReportsModule } from './reports/reports.module';
 
 /** Main Routing Module */
 import { AppRoutingModule } from './app-routing.module';
+import { LoansModule } from './loans/loans.module';
```
Comment (muskankhedia) [inline src/app/app.module.ts:42]: Please revert this change.

File: src/app/app.module.ts [hunk 14]
```diff
@@ -55,6 +56,7 @@ import { AppRoutingModule } from './app-routing.module';
     CoreModule,
     HomeModule,
     LoginModule,
+    LoansModule,
```
Comment (muskankhedia) [inline src/app/app.module.ts:59]: Please revert this change.

File: src/app/loans/loans-view/loan-account-actions/add-collateral/add-collateral.component.ts [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+   * @param {ActivatedRoute} route Activated Route.
+   * @param {LoansService} LoansService loans service.
+   */
+  constructor(private formBuilder: FormBuilder,
+              private router: Router,
+              private route: ActivatedRoute,
+              private loanService: LoansService ) { }
+
+  ngOnInit() {
+    this.setCollateralForm();
```
Comment (muskankhedia) [inline src/app/loans/loans-view/loan-account-actions/add-collateral/add-collateral.component.ts:39]: Can you please change the name to `createAddCollateralForm()`, as this may create confusion.

File: src/app/loans/loans-view/loan-account-actions/add-collateral/add-collateral.component.ts [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+   * Submits collateral form.
+   */
+  submit() {
+    const collateralTypeId = this.collateralForm.value.collateralTypeId;
+    this.collateralForm.patchValue({
+      'collateralTypeId': collateralTypeId.toString()
+    });
+    const loanId = this.route.parent.snapshot.params['loanId'];
+    const collateralForm = this.collateralForm.value;
+    collateralForm.dateFormat = 'yyyy-MM-dd';
```
Comment (muskankhedia) [inline src/app/loans/loans-view/loan-account-actions/add-collateral/add-collateral.component.ts:63]: `dateFormat` parameter isn't required to add collateral data.

File: src/app/loans/loans-view/loan-account-actions/add-collateral/add-collateral.component.ts [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  /**
+   * Submits collateral form.
+   */
+  submit() {
+    const collateralTypeId = this.collateralForm.value.collateralTypeId;
+    this.collateralForm.patchValue({
+      'collateralTypeId': collateralTypeId.toString()
+    });
```
Comment (muskankhedia) [inline src/app/loans/loans-view/loan-account-actions/add-collateral/add-collateral.component.ts:60]: The `collateralTypeId` has to be number type, so you don't need to parse it to String type.

**Comment by muskankhedia** [comment]
@BLasan Also update the PR title and the commit message as `feat: add loan action (add collateral)`

**Comment by BLasan** [comment]
@muskankhedia I'll resolve these asap

File: src/app/shared/shared.module.ts (hunk 12 above)
Comment (BLasan) [reply src/app/shared/shared.module.ts:32]: Remove this?

**Comment by BLasan** [comment]
@muskankhedia Done. Please review

File: src/app/shared/shared.module.ts [hunk 18]
```diff
@@ -28,8 +28,7 @@ import { MaterialModule } from './material.module';
   imports: [
     CommonModule,
     IconsModule,
-    MaterialModule,
-    ReactiveFormsModule
+    MaterialModule
```
Comment (muskankhedia) [inline src/app/shared/shared.module.ts:31]: Please revert all the changes made in this file.

Comment (BLasan) [reply src/app/shared/shared.module.ts:31]: Material module and reactiveforms module?

Comment (BLasan) [reply src/app/shared/shared.module.ts]: Adjusted as in the master branch shared.module.ts

**Comment by muskankhedia** [comment]
@BLasan Please fix the conflicts in the PR.

**Comment by BLasan** [comment]
@muskankhedia  Done
